src/Nethermind/Nethermind.AuRa/Validators/ContractValidator.cs

```diff
@@ -162,7 +157,31 @@ public override void PreProcess(Block block, ProcessingOptions options = Process
             
             _lastProcessedBlockNumber = block.Number;
         }
-        
+
+        private PendingValidators TryGetInitChangeFromPastBlocks(Keccak blockHash)
+        {
+            PendingValidators pendingValidators = null;
+            var lastFinalized = _blockFinalizationManager.GetLastLevelFinalizedBy(blockHash);
+            var toBlock = Math.Max(lastFinalized, InitBlockNumber);
+            var block = _blockTree.FindBlock(blockHash, BlockTreeLookupOptions.None);
+            while (block?.Number >= toBlock)
+            {
+                var receipts = block.Transactions.Select(t => _receiptStorage.Find(t.Hash)).Where(r => r != null).ToArray();
```

Comment:
how often is that called? perf wise looks like a mini-massacre

Reply:
if invoked only once then ok, if often then definitely looks scary

Reply:
Only on switch from fast to full sync